src/ILCompiler.ReadyToRun/src/ObjectWriter/R2RPEBuilder.cs

```diff
@@ -328,34 +355,18 @@ protected override BlobBuilder SerializeSection(string name, SectionLocation loc
                         // the blob data goes out of sync and WriteContentTo outputs garbage.
                         sectionDataBuilder = PEResourceHelper.Relocate(inputSectionReader, rvaDelta);
                     }
-                    else if (name == ".text")
-                    {
-                        int inputMetadataRva = _peReader.PEHeaders.CorHeader.MetadataDirectory.RelativeVirtualAddress;
-                        int inputMetadataSectionRva = sectionHeader.VirtualAddress;
-
-                        // The metadata is in .text always ... right?
-                        Debug.Assert(inputMetadataRva >= sectionHeader.VirtualAddress && inputMetadataRva < (sectionHeader.VirtualAddress + sectionHeader.VirtualSize));
-
-                        sectionDataBuilder = new BlobBuilder();
-
-                        // Write section contents up to the start of metadata
-                        sectionDataBuilder.WriteBytes(inputSectionReader.CurrentPointer, inputMetadataRva - sectionHeader.VirtualAddress);
-
-                        // Fix-up RVAs in the metadata stream 
-                        MetadataRvaFixupBuilder.Relocate(_peReader, rvaDelta).WriteContentTo(sectionDataBuilder);
-
-                        // Write the rest of the section after the metadata
-                        inputSectionReader.Offset = inputMetadataRva + _peReader.PEHeaders.MetadataSize - sectionHeader.VirtualAddress;
-                        sectionDataBuilder.WriteBytes(inputSectionReader.CurrentPointer, bytesToRead - inputSectionReader.Offset);
-
-                        Debug.Assert(sectionDataBuilder.Count == inputSectionReader.Length);
-                    }
                     else
                     {
                         sectionDataBuilder = new BlobBuilder();
                         sectionDataBuilder.WriteBytes(inputSectionReader.CurrentPointer, inputSectionReader.RemainingBytes);
                     }
 
+                    int metadataRvaDelta = _peReader.PEHeaders.CorHeader.MetadataDirectory.RelativeVirtualAddress - sectionHeader.VirtualAddress;
+                    if (metadataRvaDelta >= 0 && metadataRvaDelta < bytesToRead)
```

Comment:
Should we add an else clause with a warning, error or assertion?

Reply:
This method runs for each section in the PE image, so this check is selective and looking for the correct section. We assert that the metadata blob was found in `RelocateMetadataBlob`.